docs/proposals/resource-quota.md

```diff
@@ -0,0 +1,404 @@
+
+The administrator is not able to filter pods by quality of service.  To
+facilitate this ability, the `Pod` should expose a new field on it's `PodSpec`
+that corresponds to its `QualityOfService` that is set by the server at
+creation and used by `ResourceQuota`.  The administrator would then
+be able to filter quota by tracking `spec.qualityOfService` to quota
+`BestEffort` pods differently than `Burstable` and `Guaranteed`.
+
+## Implementation Phases and Anticipated Issues
+
+Phase 1:
+
+* Support ability for cluster-admin to quota based on workload duration
+ * Add `spec.restartAlways` as a field selector for `Pod`
+ * Ensure API server support for set-based and equality based matching of field selector
+ * Add `FieldSelector` to `v1` API
+ * Add `FieldSelector` to `ResourceQuotaSpec` to enable tracking on restart policy
+ * Add logic to admission control and quota controller to enable support
+
+* Improve latency of quota controller observing freed resources for non-pod types
+by watching for more resource types
+
+* Identify and solve issues when running large numbers of namespaces and quotas
+ * Target 20k namespaces with 60k quota documents as baseline scenario
+ * Minimize contention on locking of quota documents at admission
+
+Phase 2:
+
+* Support ability for cluster-admin to quota based on quality of service
+ * Add new field to pod that denotes quality of service class
+ * Ensure field is populated as part of `ResourceQuota` admission plug-in based on
+ enumerated resource requirements during create
+ * For operators that do not run `ResourceQuota`, ensure field is calculated in
+ pod strategy
+
+Phase 3:
+
+* Support ability to define replenishment policy to control rate of creation
+  * Add `ReplenishmentPolicy` to `ResourceQuota`
+
+* Measure performance impact of various time-based replenishment policy intervals
+with large numbers of `ResourceQuota` documents (~60k) in a cluster
+  * Target `Daily` replenishment policy as first option
+
+* Minimize controllers that operate on other users behalf using up a user's
+creation rate.  For example, a replication controller that keeps trying to
+scale up a resource may deplete the user's daily creation rate.  Controller's
+should be sensitive to quota by not making requests that would just be rejected
+by quota.  For example, if a quota allowed 10 pods in total, but the user set
+a replication controller replicaSet to 100, and the admin had a daily creation
+rate of 1000, the controller would rapidly deplete the creation rate.  If we
+target a creation rate to a particular service account or user-identity that is
+separate from the controllers, better results could potentially be achieved.
+
+Phase 4:
+
+* Support ability for namespace-admin to quota based on local understanding of
+workload
+ * Add `LabelSelector` to `v1` API
+ * Add `LabelSelector` to `ResourceQuotaSpec` to enable scoping on matched labels
+
+One of the interesting challenges with support a `LabelSelector` based
```

Comment:
Should you leave `LabelSelector` out of `ResourceQuotaSpec` until you figure this out? Otherwise it seems you would need a concept of immutable labels keys (which I'd actually favor anyway, and is something MetadataPolicy #18262 is supposed to enable) if you don't want to have to deal with objects moving in and out of different quota buckets during their lifetime due to users changing label values.